lib/Segment/QueueConsumer.php

```diff
@@ -119,11 +157,62 @@ public function flush() {
 
     $count = count($this->queue);
     $success = true;
+    $batch_size = $this->batch_size;
 
     while($count > 0 && $success) {
 
-      $batch = array_splice($this->queue, 0, min($this->batch_size, $count));
-      $success = $this->flushBatch($batch);
+        /*
+         * We use array_slice to non destructively get a batch from the queued
+         * array of calls. In the event that the batch is too big, we will try
+         * with a smaller batch size.
+         */
+        $batch = array_slice($this->queue, 0, min($batch_size, $count));
+
+        $body = $this->payload($batch);
+        $payload = json_encode($body);
+
+        /**
+         * There is a maximum of 500kb per batch request and 32kb per call.
+         *
+         * If checking for the max request size is enabled, we'll want to ensure
+         * that the entire payload is less than 500kb.
+         *
+         * If it is over 500kb, we retry at a smaller batch size until it's under
+         * the limit.
+         *
+         * @see https://segment.com/docs/sources/server/http/#max-request-size
+         */
+        if ($this->check_max_request_size) {
+
+            if ($payload_size = strlen($payload) >= self::MAX_REQUEST_BATCH_SIZE_LIMIT ) {
+
+                /*
+                 * If the payload size is too large with only 1 call in it then
+                 * there isn't anything else we can do but report it as an error.
+                 */
+                if ($batch_size == 1) {
+
+                    $this->handleError(400, "The batch size is over the API limit ($payload_size bytes)");
+
+                    return false;
+                }
+
+                /*
+                 * Reduce the batch size and try again
+                 */
+                $batch_size--;
```

Comment:
I'd make this reduce the batch size by half, instead of decrementing.. seems like a faster way to try something that should work?  Another way to do it might be to look at the size and number of calls, and assume each call is the same size and reduce the batch_size to the number of calls that would get it under the 500kb limit.